analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs

```diff
@@ -77,6 +78,9 @@ private static SymbolicConstraint ConstraintFromConstantPattern(ProgramState sta
             return null;
         }
 
+        private static SymbolicConstraint ConstraintFromDeclarationPattern(IDeclarationPatternOperationWrapper declaration, bool useOpposite) =>
+            declaration.MatchesNull || !declaration.MatchedType.IsReferenceType ? null : ObjectConstraint.NotNull.ApplyOpposite(useOpposite);
```

Comment:
I think we also missed this for the constant pattern:
```cs
_ = someObj is 5; // someObj is not null in the match case
// Also
_ = someObj is true; // someObj is not null and true in the match case
```
This applies to `dynamic` object s as well.

Reply:
It's actually both, matched type and input type. If any of them is not reference, than it doesn't make sense to build the object constraint.

We can't go into line 45 because of constant pattern and bool values.

Reply:
> It's actually both, matched type and input type. If any of them is not reference, than it doesn't make sense to build the object constraint.

If the input type is a reference type, we know in the "true" branch that it is not null.